strings/duval.cpp

```diff
@@ -0,0 +1,98 @@
+/**
+* @file duval.cpp
+ * @brief Implementation of [Duval's algorithm](https://en.wikipedia.org/wiki/Lyndon_word).
+ *
+ * @details
+ * Duval's algorithm is an algorithm to find the lexicographically smallest
+ * rotation of a string. It is based on the concept of Lyndon words.
+ * Lyndon words are defined as the lexicographically smallest string in a
+ * rotation equivalence class. A rotation equivalence class is a set of strings
+ * that can be obtained by rotating a string. For example, the rotation
+ * equivalence class of "abc" is {"abc", "bca", "cab"}. The lexicographically
+ * smallest string in this class is "abc".
+ *
+ * Duval's algorithm works by iterating over the string and finding the
+ * smallest rotation of the string that is a Lyndon word. This is done by
+ * comparing the string with its suffixes and finding the smallest suffix that
+ * is lexicographically smaller than the string. This suffix is then added to
+ * the result and the process is repeated with the remaining string.
+ * The algorithm has a time complexity of O(n) where n is the length of the
+ * string.
+ *
+ * @note While Lyndon words are described in the context of strings,
+ * Duval's algorithm can be used to find the lexicographically smallest cyclic
+ * shift of any sequence of comparable elements.
+ *
+ * @author [Amine Ghoussaini](https://github.com/aminegh20)
+*/
+
+#include <array>    ///< for std::array
+#include <cassert>  ///< for assert
+#include <cstddef>  ///< for std::size_t
+#include <deque>    ///< for std::deque
+#include <iostream> ///< for std::cout and std::endl
+#include <string>   ///< for std::string
+#include <vector>   ///< for std::vector
+
+namespace string {
+/**
+ * @brief Find the lexicographically smallest cyclic shift of a sequence.
+ * @tparam T type of the sequence
+ * @param s the sequence
+ * @returns the 0-indexed position of the least cyclic shift of the sequence
+ */
+template <typename T>
```

Comment:
Unsure if this needs to be this generic
Why not just a basic string? std::basic_string? The algorithm is entirely about strings right?

Reply:
The point here is that you could pass in an array of integers as well. In theory, this algorithm is is studied in the context of strings, but the idea here is that you do not need to have "chars". In fact, you could even pass in an array of custom comparable objects.